lib/apollo/withData.js

```diff
@@ -1,85 +1,87 @@
 import React from 'react'
-import PropTypes from 'prop-types'
-import { ApolloProvider, getDataFromTree } from 'react-apollo'
-import initApollo from './initApollo'
+import initApolloClient from './initApollo'
+import Head from 'next/head'
+import { getDataFromTree } from 'react-apollo'
 
-// Gets the display name of a JSX component for dev tools
-function getComponentDisplayName (Component) {
-  return Component.displayName || Component.name || 'Unknown'
-}
+export default App => {
+  return class WithApolloClient extends React.Component {
+    static async getInitialProps (pageContext) {
+      const { Component, router } = pageContext
 
-export default ComposedComponent => {
-  class WithData extends React.Component {
-    static async getInitialProps (ctx) {
-      let serverState = {}
-      // Evaluate the composed component's getInitialProps()
-      let composedInitialProps = {}
-      if (ComposedComponent.getInitialProps) {
-        composedInitialProps = await ComposedComponent.getInitialProps(ctx)
+      let appProps = {}
+      if (App.getInitialProps) {
+        appProps = await App.getInitialProps(
+          pageContext
+        )
       }
 
-      // we forward the accept header for webp detection
-      // - never forward cookie to client!
       const headers = !process.browser
-        ? {accept: ctx.req.headers.accept}
+        ? pageContext.ctx.req.headers
         : {}
 
+      const apolloState = {}
+
       // Run all GraphQL queries in the component tree
       // and extract the resulting data
-      if (!process.browser) {
-        // server-side we initialize apollo with all headers (including cookie)
-        const apollo = initApollo(undefined, ctx.req.headers)
-        // Provide the `url` prop data in case a GraphQL query uses it
-        const url = {
-          query: ctx.query,
-          pathname: ctx.pathname,
-          asPath: ctx.asPath
-        }
-
-        try {
-          // Run all GraphQL queries
-          await getDataFromTree(
-            <ApolloProvider client={apollo}>
-              <ComposedComponent url={url} {...composedInitialProps} serverContext={ctx} headers={headers} />
-            </ApolloProvider>
-          )
-        } catch (error) {
-          console.log(error)
-          // Prevent Apollo Client GraphQL errors from crashing SSR.
-          // Handle them in components via the data.error prop:
-          // http://dev.apollodata.com/react/api-queries.html#graphql-query-data-error
-        }
+      const apollo = initApolloClient(
+        undefined,
+        (!process.browser && headers) || undefined
```

Comment:
on the server you should use the raw headers object with cookies: `pageContext.req.headers` (fwd to backend, on client cookies are managed and sent by the browser)